### PR TITLE
Fix keybindings for Next/Previous Difference on macOS

### PR DESCRIPTION
**Pull request by krlmlr**
Ctrl + Up/Down is typically a system hotkey.

Can we also change the hotkey for switching tabs to Ctrl + (Shift + ) Tab? Page Up/Down isn't easy to access.

### CODE REVIEW

**Comment by MightyCreak** [comment]
Well, I remember a few years ago I proposed the GNOME team to change their HIG in order to use <kbd>Ctrl+Tab</kbd> and <kbd>Shift+Ctrl+Tab</kbd> to switch between tabs (preferably in a Most Recently Used -- MRU -- fashion). It was listing <kbd>Ctrl+PageUp</kbd> and <kbd>Ctrl+PageDown</kbd> which, I agree, is not very practical, and is even worse if we want to implement an MRU behaviour.

My proposition was merged and then reverted because <kbd>Ctrl+Tab</kbd> was used elsewhere for accessibility purpose, so they couldn't change that. I argued that <kbd>Ctrl+Tab</kbd> was a well known shortcut for tabs and we might consider changing the accessibility shortcut instead, but it didn't get much traction (you don't mess with accessibility I guess :sweat_smile: ).

All that said, I checked again and they seem to have loosen the guidelines a bit:
* [Tabs]((https://developer.gnome.org/hig/patterns/nav/tabs.html?highlight=ctrl%2btab)): <kbd>Ctrl+Tab</kbd> is just "Switch tabs" now, but they don't talk about <kbd>Shift+Ctrl+Tab</kbd> 
* [Keyboard](https://developer.gnome.org/hig/guidelines/keyboard.html?highlight=ctrl%20tab): they allow a different behaviour between <kbd>Tab</kbd>/<kbd>Shift+Tab</kbd> and <kbd>Ctrl+Tab</kbd>/<kbd>Shift+Ctrl+Tab</kbd>

So it seems that they are more inclined to let the devs choose whatever behaviour they want. They are still not talking about MRU, but that's another issue.

The thing though is that it's for GTK4, and I don't know how to change that in GTK3 as the tab widget is simply following the GTK3 HIG. It seems it will require some custom overloading of the widget.

I also know that they did some standardisation around all the tabs widget in GTK4 as all the devs were doing their own behaviours depending on the apps.

TL;DR: it's not easy to change in GTK3 apparently (or at least I'm not expert enough yet to do that), and GTK4 seems to have simplified the tabs widget.

**Comment by krlmlr** [comment]
I'm confused, don't we specify the hotkeys in our code?

https://github.com/MightyCreak/diffuse/blob/122879df07ae316298073d83ed7df2b9ee1b0ef1/src/diffuse/resources.py#L82-L85

**Comment by krlmlr** [comment]
(Please, if we do MRU, can it be an option, and implement easy tab relocation instead? And perhaps accessing individual tabs with a hotkey?)

**Comment by MightyCreak** [comment]
> I'm confused, don't we specify the hotkeys in our code?
> 
> https://github.com/MightyCreak/diffuse/blob/122879df07ae316298073d83ed7df2b9ee1b0ef1/src/diffuse/resources.py#L82-L85

Good point, I thought Diffuse was relaying on GTK's widget for that. It's worth trying at least. At least changing it for Ctrl+Tab and Shift+Ctrl+Tab and see how it goes.

> (Please, if we do MRU, can it be an option, and implement easy tab relocation instead? And perhaps accessing individual tabs with a hotkey?)

Won't be done right now, and can surely be an option, don't worry :wink: 

Easy tab relocation? You mean reordering the tabs manually?

We could have something like Alt+1/2/3/4/5 to quickly access the tabs (as in Firefox for instance).

**Comment by krlmlr** [comment]
Yes, reordering tabs manually. Alt+<number> or Apple+<number> (on the Mac) would be good to have too. I'll try in a separate PR.

Side note: the color scheme for dark mode isn't ideal yet on my system. I remember tweaking it in a previous installation.

**Comment by MightyCreak** [comment]
> Side note: the color scheme for dark mode isn't ideal yet on my system. I remember tweaking it in a previous installation.

I was looking at that when I dabble on the Rust syntax earlier. I think it would be possible to create a common color scheme for all the languages, instead of setting the comment color for each one for instance.

I was thinking of something like `var <var_name> "<var_value>"`, and then use the `colour` command like that: `colour <keyword> @<var_name>` and it would simply replace `@<var_name>` with it's value.

Then we could have a dark and a light schemes for all languages (or even more schemes if we want).

TBC

File: src/diffuse/resources.py [hunk 1]
```diff
@@ -75,10 +75,10 @@ def __init__(self):
         self.setKeyBinding('menu', 'go_to_line', defaultModKey + 'Shift+l')
         self.setKeyBinding('menu', 'realign_all', defaultModKey + 'l')
         self.setKeyBinding('menu', 'isolate', defaultModKey + 'i')
-        self.setKeyBinding('menu', 'first_difference', 'Shift+Ctrl+Up')
-        self.setKeyBinding('menu', 'previous_difference', 'Ctrl+Up')
-        self.setKeyBinding('menu', 'next_difference', 'Ctrl+Down')
-        self.setKeyBinding('menu', 'last_difference', 'Shift+Ctrl+Down')
+        self.setKeyBinding('menu', 'first_difference', defaultModKey + 'Shift+Up')
+        self.setKeyBinding('menu', 'previous_difference', defaultModKey + 'Up')
```
Comment (hugoholgersson) [inline src/diffuse/resources.py:79]: > Ctrl + Up/Down is typically a system hotkey.

What in macOS is mapped to Ctrl + Up / Down? I could not see any conflicts myself.

I'd rather keep these 4 commands to the Ctrl button. When I created `defaultModKey`, I did not append it to these 4 bindings because, also on macOS, I felt it was more natural to use the Ctrl-key for everything "moving between differences"... Besides, just because we are on macOS, that doesn't mean we need to use the Command key for everything?

Compare with Visual Studio Code on macOS, for example. Its default keybindings ⌘↑ / ⌘↓ means "Go to beginning/end of file" but nothing is mapped to Ctrl + ↑ / ↓.

Source: https://code.visualstudio.com/shortcuts/keyboard-shortcuts-macos.pdf

**Review by hugoholgersson — CHANGES_REQUESTED** [review]
I'd rather see that we don't land this. (Or I will need to rewire my muscle memory...)

File: src/diffuse/resources.py (hunk 1 above)
Comment (krlmlr) [reply src/diffuse/resources.py:79]: For "Mission Control" on my fairly recent M2 system:

image

What OS version are you on?

I don't have a strong opinion about those shortcuts (and I can run a fork anyway), but other users with default settings will hit this exact problem.

Comment (krlmlr) [reply src/diffuse/resources.py:79]: https://support.apple.com/guide/mac-help/open-windows-spaces-mission-control-mh35798/mac

Comment (hugoholgersson) [reply src/diffuse/resources.py:79]: I see. Then it makes sense to use ⌘ instead, I guess.

**Comment by MightyCreak** [comment]
I just understood that we were talking about the previous/next **difference**. I was sure you were talking about tabs there.

That changes a few things about what I said, which must have been pretty out of context :sweat_smile: 

Anyway, you're the Mac guys, I'll let you decide what would be the best. But there must be a good rational. Like, for instance, how other diff tools are doing it? Or why Ctrl+Up/Down is very bad on Mac?

Changing shortcuts needs a lot of reflection because it will frustrate a lot of people because of muscle memory.

File: src/diffuse/resources.py (hunk 1 above)
Comment (MightyCreak) [reply src/diffuse/resources.py:79]: I'm willing to change <kbd>Ctrl</kbd> to <kbd>⌘</kbd> for the first/previous/next/last difference.

I do wonder though why not all the <kbd>Ctrl</kbd> here are not changed to <kbd>⌘</kbd>? It is my understanding that Mac doesn't really use <kbd>Ctrl</kbd>...

As Mac users, what is your experience with these bindings? The fact that some are using <kbd>Ctrl</kbd> and some are using <kbd>⌘</kbd>?

**Review by MightyCreak — APPROVED** [review]

File: src/diffuse/resources.py (hunk 1 above)
Comment (krlmlr) [reply src/diffuse/resources.py:79]: Ctrl can be used, but perhaps mostly as a modifier to Cmd, which is the key that's used most frequently for commands.

Happy to address this in a new PR?

Comment (MightyCreak) [reply src/diffuse/resources.py:79]: Yep, I'll let you and @hugoholgersson (our two official Mac people on the project :wink: ) determine what would be the best modifiers to use on Mac.
